File: internal/webui/static/assets/detector-coverage.js
// detector-coverage.js — pure presentation helpers for the daemon-coverage
// matrix on the Detectors page (GET /api/v1/detectors/coverage). Verdicts are
// host-reality aware: "absent" means the daemon AND its detector are both
// missing, which is informational rather than a problem — it must NOT render
// as an error.

const VERDICT_BY_SUMMARY_KEY = {
  ok: 'ok',
  gaps: 'gap',
  disabled: 'disabled',
  dormant: 'dormant',
  absent: 'absent',
  event_driven: 'na',
};

export function coverageBadge(verdict) {
  switch (String(verdict || '').toLowerCase()) {
    case 'ok':
      return { cls: 'pill ok', label: 'OK', title: 'Daemon running and detector enabled' };
    case 'gap':
      return { cls: 'pill danger', label: 'GAP', title: 'Daemon is running but no section enables this detector' };
    case 'disabled':
      return { cls: 'pill warn', label: 'disabled', title: 'Daemon running, but every section has ENABLED=0' };
    case 'dormant':
      return { cls: 'pill warn', label: 'dormant', title: 'Detector enabled but its daemon is absent or stopped' };
    case 'absent':
      return { cls: 'pill', label: 'absent', title: 'Daemon not on this host and detector not configured — informational' };
    case 'na':
      return { cls: 'pill info', label: 'n/a', title: 'Event-driven detector (no daemon)' };
    default:
      return { cls: 'pill', label: String(verdict || '?'), title: '' };
  }
}

export function coverageSummaryItems(summary) {
  const s = summary || {};
  return Object.entries(VERDICT_BY_SUMMARY_KEY).map(([key, verdict]) => {
    const badge = coverageBadge(verdict);
    const count = Number(s[key] || 0);
    return { key, count, cls: badge.cls, title: badge.title, label: `${badge.label}: ${count}` };
  });
}

export function unitStateText(unit) {
  const u = unit || {};
  if (!u.found) return `${u.unit || '?'}: not installed`;
  if (!u.active) return `${u.unit || '?'}: stopped`;
  return `${u.unit || '?'}: running`;
}

File: internal/mcpserver/detector_coverage.go
package mcpserver

// detector_coverage.go — read-only MCP tool over GET /api/v1/detectors/coverage:
// the daemon-vs-detector coverage matrix. For every registered detector type it
// answers whether the watched daemon exists on this host (svcstat over a curated
// type→units affinity table), whether a live detectors.conf section enables it,
// and a reality-aware verdict so "not configured" stops being noise when the
// daemon itself is absent.

import (
	"context"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

func registerDetectorCoverage(srv *mcp.Server, d Deps) {
	mcp.AddTool(srv, &mcp.Tool{
		Annotations: readOnly,
		Name:        "detector_coverage",
		Description: "Daemon-vs-detector coverage matrix for this node — 'what runs here that nothing watches, and what watches something that isn't there?'. For every registered detector type (ssh_auth, exim_*/postfix_* queues+security+relays, dovecot_auth, ftpd, mysql(+governor), cpanel, modsec, proxmox_auth, webdetector, …) it probes whether the watched daemon's systemd unit EXISTS and is ACTIVE (curated affinity: sshd/exim/postfix/dovecot/vsftpd|pure-ftpd|proftpd/mysqld|mariadb/cpanel/modsec-inside-webserver/pvedaemon/edge), lists the live detectors.conf sections for it (configured / ENABLED / active / log-source-probe-ok) and returns a verdict: ok; GAP (daemon RUNNING but no section or all ENABLED=0 — the real 'I forgot' case); disabled; dormant (enabled but daemon absent/stopped — it will idle or fail its source probe); absent (daemon not present AND not configured — informational, NOT a problem); na (event-driven, no daemon). Verdicts are host-reality-aware, so absent types are not complaints. Start from summary.gaps + summary.dormant, then read the per-type rows; sections[] carries runtime detail (last_error, init diagnostics). Pair with detectors_status for run counters.",
	}, func(ctx context.Context, _ *mcp.CallToolRequest, _ emptyInput) (*mcp.CallToolResult, any, error) {
		return dispatchJSON(ctx, d, "/api/v1/detectors/coverage", nil)
	})
}
